### PR TITLE
fix #170

### LINKED ISSUE #170: When using TrimZero, size do not match error occur
When using `TrimZero` encounter an error like below:
```
/usr/local/torch/install/bin/luajit: /usr/local/torch/install/share/lua/5.1/rnn/Dropout.lua:43: bad argument #2 to 'cmul' (sizes do not match at /tmp/luarocks_cutorch-scm-1-9715/cutorch/lib/THC/THCTensorMathPointwise.cu:160)
stack traceback:
	[C]: in function 'cmul'
	/usr/local/torch/install/share/lua/5.1/rnn/Dropout.lua:43: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/ConcatTable.lua:11: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/ParallelTable.lua:12: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/ConcatTable.lua:11: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/rnn/TrimZero.lua:120: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/rnn/GRU.lua:152: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/rnn/Sequencer.lua:48: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/ParallelTable.lua:12: in function 'updateOutput'
	/usr/local/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	train_residual.lua:295: in function 'opfunc'
	/usr/local/torch/install/share/lua/5.1/optim/rmsprop.lua:32: in function 'rmsprop'
	train_residual.lua:335: in main chunk
	[C]: in function 'dofile'
	...ocal/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00406670
```

### PR DESCRIPTION
**Pull request by jnhwkim**
I discovered there was an interaction between TrimZero and BGRU after polishing the trimZero code.

Lazy Dropout couldn't handle variable input (batch) size after trimming, I added a `mono` option to sample once for a batch for expanding across the batch.

It fixed #170.

### CODE REVIEW

File: AbstractRecurrent.lua [hunk 1]
```diff
@@ -40,7 +41,8 @@ end
 
 function AbstractRecurrent:trimZero(nInputDim)
    self.recurrentModule = nn.TrimZero(self.recurrentModule, nInputDim, true)
-   self.sharedClones = {self.recurrentModule}
+   self.sharedClones = {self.recurrentModule:stepClone()}
```
Comment (nicholas-leonard) [inline AbstractRecurrent.lua]: This should be `self.sharedClones = {self.recurrentModule}` or it will cause issues with BatchNormalization and such.

Comment (jnhwkim) [inline AbstractRecurrent.lua]: @nicholas-leonard please check my new pull request.

File: test/test.lua [hunk 2]
```diff
@@ -3826,6 +3826,56 @@ function rnntest.TrimZero()
          end
       end
    end
+
+   -- check to have the same loss
```
Comment (nicholas-leonard) [inline test/test.lua]: Can you make these `local` variables?

File: test/test.lua [hunk 3]
```diff
@@ -4192,6 +4242,54 @@ function rnntest.issue129()
    mytester:assertTensorEq(output, output2,  0.0002, "issue 129 err")
 end
 
+function rnntest.issue170()
+   torch.manualSeed(123)
+
+   rnn_size = 8
```
Comment (nicholas-leonard) [inline test/test.lua]: Same. `local`.

File: test/test.lua (hunk 2 above)
Comment (jnhwkim) [inline test/test.lua]: @nicholas-leonard done

**Comment by jnhwkim** [comment]
@nicholas-leonard please check my new PR!

File: test/test.lua (hunk 3 above)
Comment (jnhwkim) [inline test/test.lua]: @nicholas-leonard done

**Comment by nicholas-leonard** [comment]
@jnhwkim Looks great. Thanks for the fixes!

**Pull request merged**